LookAndFeels/foleys_Skeuomorphic.cpp

```diff
@@ -0,0 +1,240 @@
+static inline void fillEllipse (Graphics& g, const Rectangle<float>& bounds, Colour colour)
+{
+    g.setColour (colour);
+    g.fillEllipse (bounds);
+}
+
+static inline void fillEllipse (Graphics& g, const Rectangle<float>& bounds, const ColourGradient& gradient)
+{
+    g.setGradientFill (gradient);
+    g.fillEllipse (bounds);
+}
+
+const Skeuomorphic::knobImages& Skeuomorphic::getKnobImages (int diameter)
+{
+    auto search = knobsBgs.find (diameter);
+    if (search != knobsBgs.end()) return search->second;
+
+    float radius = diameter * 0.5f;
+    bool isTiny  = diameter < 20;
+    bool isSmall = diameter < 60;
+
+    Point<float>     centre { radius, radius };
+    Rectangle<float> bounds { 0, 0, float (diameter), float (diameter) };
+
+    if (knobsBgs.size() > size_t (knobsMaxSize))
+        knobsBgs.clear();
+
+    Image bg = Image (Image::ARGB, diameter, diameter + 3, true);
+
+    Graphics g (bg);
+    fillEllipse (g, bounds.translated (0.0f, 1.0f), whiteA010);
+    fillEllipse (g, bounds.translated (0.0f, 2.0f), whiteA010);
+    fillEllipse (g, bounds,                         { 50, 51, 61});
+
+    auto backgroundGr = isTiny? makeGradient (centre, { centre.x, centre.y + radius }, true,
+                                          {  { 0.0, Colours::transparentBlack },
+                                             { 0.9, Colours::transparentBlack },
+                                             { 1.0, blackA122 }  }) :
+                                makeGradient (centre, { centre.x, centre.y + radius }, true,
+                                          {  { 0.0,                          Colours::transparentBlack },
+                                             { 0.52,                         blackA092 },
+                                             { 0.9,                          Colours::transparentBlack },
+                                             { double (1 - (4.0f / radius)), Colours::transparentBlack },
+                                             { 1.0,                          blackA092 } });
+    fillEllipse (g, bounds, backgroundGr);
+
+
+
+    bounds.reduce (diameter * 0.15f, diameter * 0.15f);
+    float fgRadius = bounds.getWidth() * 0.5f;
+
+    Image fg = Image (Image::ARGB, diameter, diameter, true);
+
+    Graphics gg (fg); // lower shadow
+    {
+        float xOffset = fgRadius * 0.12f;
+        float yOffset = fgRadius * 0.32f;
+
+        ColourGradient shadowGr = makeGradient ({ centre.x + xOffset, centre.y + yOffset },
+                                                { centre.x, centre.y + fgRadius + yOffset }, true,
+                                             {  { 0.0,  blackA122                 },
+                                                { 0.72, blackA122                 },
+                                                { 1.0,  Colours::transparentBlack }  });
+        fillEllipse (gg, bounds.translated (xOffset, yOffset), shadowGr);
+    }
+
+    // middle
+    {
+        gg.saveState();
```

Comment:
better to use:
```
{
    juce::Graphics::ScopedSaveState save (gg);
    // ...
}
// end of scope makes sure, there is no saveState/restoreState imbalance
```

Reply:
I'm actually calling `gg.restoreState();` in the same scope a few lines later before drawing this middle part's outline. I'm using curly brackets to make clear what parts of the code are drawing what pieces of the knob.